### PR TITLE
feat: merge dev xiujuan own commit pr to main

### PR DESCRIPTION
**Pull request by o0OooO**
## Description

Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

**Comment by NingLu** [comment]
We identified new issues on unchanged lines of code. Navigate to the Amazon CodeGuru Reviewer console to view the recommendations to fix them.

File: aws_extension/mme_utils.py [hunk 1]
```diff
@@ -100,11 +105,24 @@ def upload_model(model_type, model_name, model_s3_pos):
...
+    if file_type == TAR_TYPE_FILE:
+        logging.info("model type is tar")
+        os.system(f"tar xvf {tar_name}")
+        os.system(f"rm {tar_name}")
+        os.system("df -h")
+    else:
+        os.system(f"rm {tar_name}")
+        logging.info(f"model type is origin file type: {file_type}")
+        prefix_name = model_s3_pos.split('.')[0]
+        os.system(f'./tools/s5cmd cp {prefix_name}* ./models/{model_type}/')
```
Comment (NingLu) [inline aws_extension/mme_utils.py:124]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Starting a process with a shell, possible injection detected, security issue. https://bandit.readthedocs.io/en/latest/plugins/b605_start_process_with_a_shell.html

File: middleware_api/lambda/model_and_train/checkpoint_api.py [hunk 2]
```diff
@@ -57,6 +62,102 @@ def list_all_checkpoints_api(event, context):
...
+            'error': "please choose the right type from :'Stable-diffusion', 'embeddings', 'Lora', 'hypernetworks', 'Controlnet', 'VAE'"
+        }
+
+    try:
+        base_key = get_base_checkpoint_s3_key(_type, 'custom', request_id)
+        urls = event.modelUrl
+        file_names = []
+        logger.info(f"start to upload models:{urls}")
+        checkpoint_params = {}
+        if event.params is not None and len(event.params) > 0:
```
Comment (NingLu) [inline middleware_api/lambda/model_and_train/checkpoint_api.py:123]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

To check if a container or sequence (string, list, tuple) is empty, use `if not val`. Do not compare its length using `if len(val) == 0`

[Learn more](https://www.python.org/dev/peps/pep-0008/#programming-recommendations#:~:text=if%20not%20seq)

File: aws_extension/mme_utils.py [hunk 3]
```diff
@@ -100,11 +105,24 @@
...
+    command = f"file --mime-type -b ./{tar_name}"
+    file_type = subprocess.check_output(command, shell=True).decode('utf-8').strip()
+    logging.info(f"file_type is {file_type}")
+    if file_type == TAR_TYPE_FILE:
+        logging.info("model type is tar")
+        os.system(f"tar xvf {tar_name}")
+        os.system(f"rm {tar_name}")
+        os.system("df -h")
+    else:
+        os.system(f"rm {tar_name}")
```
Comment (NingLu) [inline aws_extension/mme_utils.py:121]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Starting a process with a shell, possible injection detected, security issue. https://bandit.readthedocs.io/en/latest/plugins/b605_start_process_with_a_shell.html

File: aws_extension/sagemaker_ui.py [hunk 4]
```diff
@@ -450,6 +458,37 @@ def get_model_list_by_type(model_type):
...
+        if "checkpoints" not in json_response.keys():
+            return []
+        checkpoint_dict = {}
+        for ckpt in json_response["checkpoints"]:
+            if "name" not in ckpt:
+                continue
+            if ckpt["name"] is None:
+                continue
+            create_time = ckpt['created']
+            created = datetime.fromtimestamp(create_time)
```
Comment (NingLu) [inline aws_extension/sagemaker_ui.py:480]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The naive datetime objects are treated by many datetime methods as local times, it is preferred to use aware datetimes to represent times in UTC. The recommended way to create an aware datetime object representing a specific timestamp in UTC is by passing `tzinfo` as an argument to the method.

[Learn more](https://docs.python.org/3/library/datetime.html#aware-and-naive-objects)

File: aws_extension/mme_utils.py [hunk 5]
```diff
@@ -100,11 +105,24 @@
...
-    os.system("df -h")
+    logging.info(tar_name)
+    command = f"file --mime-type -b ./{tar_name}"
+    file_type = subprocess.check_output(command, shell=True).decode('utf-8').strip()
+    logging.info(f"file_type is {file_type}")
+    if file_type == TAR_TYPE_FILE:
+        logging.info("model type is tar")
+        os.system(f"tar xvf {tar_name}")
+        os.system(f"rm {tar_name}")
+        os.system("df -h")
```
Comment (NingLu) [inline aws_extension/mme_utils.py:119]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Starting a process with a partial executable path https://bandit.readthedocs.io/en/latest/plugins/b607_start_process_with_partial_path.html

File: middleware_api/lambda/inference_result_notification/app.py [hunk 6]
```diff
@@ -228,7 +228,7 @@ def lambda_handler(event, context):
                 print(f"Complete inference parameters {inference_parameters}")
 
             update_inference_job_table(inference_id, 'status', 'succeed')
-            update_inference_job_table(inference_id, 'completeTime', get_curent_time())
+            update_inference_job_table(inference_id, 'completeTime', str(datetime.now()))
```
Comment (NingLu) [inline middleware_api/lambda/inference_result_notification/app.py:231]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The naive datetime objects are treated by many datetime methods as local times, it is preferred to use aware datetimes to represent times in UTC. The recommended way to create an aware datetime object representing a specific timestamp in UTC is by passing `tzinfo` as an argument to the method.

[Learn more](https://docs.python.org/3/library/datetime.html#aware-and-naive-objects)

File: aws_extension/mme_utils.py [hunk 7]
```diff
@@ -100,11 +105,24 @@
...
-    os.system(f"rm {tar_name}")
-    os.system("df -h")
+    logging.info(tar_name)
+    command = f"file --mime-type -b ./{tar_name}"
+    file_type = subprocess.check_output(command, shell=True).decode('utf-8').strip()
+    logging.info(f"file_type is {file_type}")
+    if file_type == TAR_TYPE_FILE:
+        logging.info("model type is tar")
+        os.system(f"tar xvf {tar_name}")
+        os.system(f"rm {tar_name}")
```
Comment (NingLu) [inline aws_extension/mme_utils.py:118]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Starting a process with a shell, possible injection detected, security issue. https://bandit.readthedocs.io/en/latest/plugins/b605_start_process_with_a_shell.html

File: aws_extension/mme_utils.py (hunk 5 above)
Comment (NingLu) [inline aws_extension/mme_utils.py:119]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Starting a process with a shell: Seems safe, but may be changed in the future, consider rewriting without shell https://bandit.readthedocs.io/en/latest/plugins/b605_start_process_with_a_shell.html

File: aws_extension/mme_utils.py [hunk 8]
```diff
@@ -100,11 +105,24 @@
...
     #download from s3
+    logging.info(f'./tools/s5cmd cp {model_s3_pos} ./')
     os.system(f'./tools/s5cmd cp {model_s3_pos} ./')
     tar_name = model_s3_pos.split('/')[-1]
-    os.system(f"tar xvf {tar_name}")
-    os.system(f"rm {tar_name}")
-    os.system("df -h")
+    logging.info(tar_name)
+    command = f"file --mime-type -b ./{tar_name}"
+    file_type = subprocess.check_output(command, shell=True).decode('utf-8').strip()
```
Comment (NingLu) [inline aws_extension/mme_utils.py:113]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

subprocess call with shell=True identified, security issue. https://bandit.readthedocs.io/en/latest/plugins/b602_subprocess_popen_with_shell_equals_true.html

File: middleware_api/lambda/model_and_train/checkpoint_api.py [hunk 9]
```diff
@@ -57,6 +62,102 @@
...
+            checkpoint_params = event.params
+        checkpoint_params['created'] = str(datetime.datetime.now())
```
Comment (NingLu) [inline middleware_api/lambda/model_and_train/checkpoint_api.py:125]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The naive datetime objects are treated by many datetime methods as local times, it is preferred to use aware datetimes to represent times in UTC. The recommended way to create an aware datetime object representing a specific timestamp in UTC is by passing `tzinfo` as an argument to the method.

[Learn more](https://docs.python.org/3/library/datetime.html#aware-and-naive-objects)

Similar issue at line number 137.

File: middleware_api/lambda/model_and_train/common_tools.py [hunk 10]
```diff
@@ -89,6 +97,73 @@ def split_s3_path(s3_path):
...
+# Returns:
+#     upload_id: s3 uploadId
+#     key:s3 key
+#     bucket: bucket name
+def multipart_upload_from_url(url, bucket_name, s3_key):
+    s3 = boto3.client('s3')
+    logging.info(f"start multipart_upload_from_url:{url}, {s3_key}")
+    lock = threading.Lock()
+    try:
+        with urllib.request.urlopen(url) as response:
```
Comment (NingLu) [inline middleware_api/lambda/model_and_train/common_tools.py:128]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Audit url open for permitted schemes. Allowing use of file:/ or custom schemes is often unexpected. https://bandit.readthedocs.io/en/latest/blacklists/blacklist_calls.html#b310-urllib-urlopen

File: aws_extension/mme_utils.py [hunk 11]
```diff
@@ -100,11 +105,24 @@
...
+    logging.info(f"file_type is {file_type}")
+    if file_type == TAR_TYPE_FILE:
+        logging.info("model type is tar")
+        os.system(f"tar xvf {tar_name}")
```
Comment (NingLu) [inline aws_extension/mme_utils.py:117]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Starting a process with a shell, possible injection detected, security issue. https://bandit.readthedocs.io/en/latest/plugins/b605_start_process_with_a_shell.html

**Review by guming3d — APPROVED** [review]

**Pull request merged**